core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java

```diff
@@ -0,0 +1,239 @@
+      );
+
+      final Enumerable<Object[]> enumerable =
+          Linq4j.asEnumerable(
+              Collections.singletonList(new Object[]{
+                  java.sql.Date.valueOf("2018-12-14"),
+                  java.sql.Time.valueOf("18:29:34"),
+                  java.sql.Timestamp.valueOf("2018-12-14 18:29:34.123")}));
+
+      final Table table = new AbstractQueryableTable(Object[].class) {
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        @Override public Queryable<Object[]> asQueryable(QueryProvider queryProvider,
+                                                         SchemaPlus schema, String tableName) {
+          return EnumerableDefaults.asOrderedQueryable(enumerable);
+        }
+
+        @Override public RelDataType getRowType(RelDataTypeFactory typeFactory) {
+          return rowType;
+        }
+      };
+
+      rootSchema.add("java_sql_date_types", table);
+
+      test.test(connection);
+    }
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-1703">[CALCITE-1703]
+   * Before the fix, functions on TIME and TIMESTAMP column can throw ClassCastException</a>. */
+  @Test public void testJavaSqlDateColumnReference() throws SQLException {
+
+    withJavaSqlDateTypes(connection -> {
+      // 'extract' function expects numeric date representation.
+      // RexToLixTranslator.convert() handles Java type to numeric storage type conversion.
+      // Storage types are defined at EnumUtils.toInternal.
+      final Statement statement = connection.createStatement();
+      String sql = "select extract(year from \"dt\"),"
+          + " extract(hour from \"tm\"),"
+          + " extract(day from \"ts\")"
+          + " from \"java_sql_date_types\"";
+
+      ResultSet resultSet = statement.executeQuery(sql);
+      assertTrue(resultSet.next());
+      assertEquals("java.sql.Date is converted to int when it's passed to a function",
+          2018, resultSet.getInt(1));
+      assertEquals("java.sql.Time is converted to int when it's passed to a function",
+          18, resultSet.getInt(2));
+      assertEquals("java.sql.Timestamp is converted to"
+              + " long when it's passed to a function",
+          14, resultSet.getInt(3));
+    });
+
+  }
+
+  /**
+   * Test to confirm java.sql.Time column can be used by a filter condition.
+   */
+  @Test public void testTimeFilter() throws SQLException {
+
+    withJavaSqlDateTypes(connection -> {
+
+      String sql = "select 'result'"
+          + " from \"java_sql_date_types\"";
+
+      Statement statement = connection.createStatement();
+      ResultSet resultSet = statement.executeQuery(sql + " where \"tm\" > TIME '18:29:33'");
+      assertTrue("Time is grater than the literal", resultSet.next());
+
+
+      statement = connection.createStatement();
+      resultSet = statement.executeQuery(sql + " where \"tm\" = TIME '18:29:34'");
+      assertTrue("Time equals to the literal", resultSet.next());
```

Comment:
Have you considered refactoring "createStatement + executeQuery + assert" to a separate method?
It looks like `"\"tm\" = TIME '18:29:34'"` could be method parameter so it is used in both `assertMessage` and query text.

Reply:
Consider that a bit, but I just updated the assertion messages. Thanks for the suggestion anyway.